### PR TITLE
[YUNIKORN-519] Cleanup placeholders when the app is Completed

### PR DESCRIPTION
**Pull request by kingamarton**

### CODE REVIEW

**Review by wilfred-s — CHANGES_REQUESTED** [review]
I think we need to check the behaviour for an application moving into Waiting and then Completed state. Waiting state can be entered in 3 cases: 
1. an ask is removed and nothing is pending (ask) or allocated (real and placeholders)  
2. an allocation is removed and nothing is pending (ask) or allocated (real and placeholders)
3. all allocations are removed and nothing is pending (ask)

Completed is only entered when we time out from waiting. So I am not sure if this is a real issue or that this will be solved indirectly by the placeholder timeout in YUNIKORN-460

**Comment by kingamarton** [comment]
> I think we need to check the behaviour for an application moving into Waiting and then Completed state. Waiting state can be entered in 3 cases:
> 
> 1. an ask is removed and nothing is pending (ask) or allocated (real and placeholders)
> 2. an allocation is removed and nothing is pending (ask) or allocated (real and placeholders)
> 3. all allocations are removed and nothing is pending (ask)
> 
> Completed is only entered when we time out from waiting. So I am not sure if this is a real issue or that this will be solved indirectly by the placeholder timeout in YUNIKORN-460

@wilfred-s, according to the [design doc](https://docs.google.com/document/d/1YpLZLqYsp5IaeEdl_AU9wZ-f16CfaaoRhZdalDP4RCg/edit):  `An application enters the waiting state at the time that there are no active allocations (allocated resources > 0) and pending allocation asks (pending resources > 0). Placeholders are not considered active allocations.`
This means that when we enter the waiting state we can still have some active placeholders, what we need to delete when the app is finished.

I think the completed state handling and the placeholder timeout are two different use cases, however both of them will perform the same cleaning. I think if an application is already completed, there is no reason for waiting until the placeholder allocation times out.

**Comment by wilfred-s** [comment]
The design doc describes a number of cases. And yes you are correct. Waiting state should not include the placeholders. I looked at the code not the design doc. Please check the application code as I have been to restrictive in the transition: L335 and L1099 do have checks for placeholders which should be removed. At that point we're back in line with the design.

But the design also shows that the clean up of the placeholders happens before leaving the Waiting state:
> For the exit out of the waiting state to the completed state the application must be clean. Processing in the core needs to consider two cases:
> 1. No placeholders allocated
> 2. Placeholders allocated

Case 1 is simple the timeout for the Waiting state triggers as per normal and no placeholders allocated sends the event.

Case 2 means that in the time out that is triggered for the waiting state should trigger the cleanup of the placeholders. This is problematic as the app cannot communicate that back to the shim. We found that while designing this solution in earlier discussions. Implementing what is in the design doc was not possible without huge implications.

So we have two things to fix up:
* add the two changes to the check in the app (remove the placeholder check and comments)
* update the design doc clean up part to show the implementation of the cleanup in the partition manager.

File: pkg/scheduler/partition.go [hunk 1]
```diff
@@ -1207,6 +1207,25 @@ func (pc *PartitionContext) cleanupExpiredApps() {
...
+		// at this point if we still have allocations those are placeholders
+		for _, alloc := range app.GetAllAllocations() {
+			r, _ := pc.removeAllocation(&si.AllocationRelease{
+				PartitionName:   pc.Name,
+				ApplicationID:   app.ApplicationID,
+				UUID:            alloc.UUID,
+				TerminationType: si.AllocationRelease_TIMEOUT,
+			})
+			released = append(released, r...)
+		}
```
[the remark this answers is not in the document]
Comment (kingamarton) [reply pkg/scheduler/partition.go:1222]: If we just simply call app.RemoveAllAllocations, then the allocations won't de deleted from the node, just from the application

[the remark this answers is not in the document]
Comment (wilfred-s) [reply pkg/scheduler/partition.go:1222]: I took a short cut, and see that I was really not clear at all.
The cascading release and cleanup can be triggered by simply passing in an empty UUID. You only do one call to `pc.removeAllocation()` using that we call `app.RemoveAllAllocations()`similar to what you do for the allocationAsks clean up

**Pull request closed without merging**